agrirouter-api-java-convenience/src/main/java/com/dke/data/agrirouter/convenience/rest/service/messaging/MessageConfirmationFunctionsService.java

```diff
@@ -68,7 +69,8 @@ private void confirmAllPendingMessages(
 
     Optional<List<FetchMessageResponse>> fetchMessageResponses =
         this.fetchMessageService.fetch(
-            parameters.getOnboardingResponse(), MAX_TRIES_BEFORE_FAILURE, DEFAULT_INTERVAL);
+            parameters.getOnboardingResponse(),
+            new DefaultCancellationToken(MAX_TRIES_BEFORE_FAILURE, DEFAULT_INTERVAL));
```

Comment:
Step 1: Add Variable to hold possible custom implementation of CancellationToken:

```
private CancellationToken cancellationToken;

  public MessageConfirmationFunctionsService(Environment environment, CancellationToken cancellationToken){
    this.messageQueryService = new MessageQueryServiceImpl(environment);
    this.fetchMessageService = new FetchMessageServiceImpl();
    this.decodeMessageService = new DecodeMessageServiceImpl();
    this.messageConfirmationService = new MessageConfirmationServiceImpl(environment);
    this.cancellationToken = cancellationToken;

  }

  public MessageConfirmationFunctionsService(Environment environment) {
    this(environment,new DefaultCancellationToken(MAX_TRIES_BEFORE_FAILURE, DEFAULT_INTERVAL));
  }
```


Step 2: Add CancellationToken here:

`    Optional<List<FetchMessageResponse>> fetchMessageResponses =
        this.fetchMessageService.fetch(
            parameters.getOnboardingResponse(),
            this.cancellationToken
            );`

Reply:
Adding the cancellation token to the constructor would mean, that you need to modify for example DI for frameworks like Spring Boot and provide a dedicated bean. This would have a big impact at this point. Since we are talking about a convenience implementation I would suggest using the default behavior.

Reply:
Same here, see above